assimilation_code/modules/observations/DEFAULT_obs_kind_mod.F90

```diff
@@ -1414,111 +638,125 @@ function get_num_quantities()
 
+call validate_obs_type_index(obs_type_ind, routine)
+
 evaluate_this_type_of_obs = obs_type_info(obs_type_ind)%evaluate
 
 end function evaluate_this_type_of_obs
 
 !----------------------------------------------------------------------------
+!> Returns true if this obs_type should use externally computed priors
 
 function use_ext_prior_this_type_of_obs(obs_type_ind)
 
-! Returns true if this obs_type should use externally computed priors
-
 logical             :: use_ext_prior_this_type_of_obs
 integer, intent(in) :: obs_type_ind
 
+character(len=*), parameter :: routine = 'use_ext_prior_this_type_of_obs'
+
 if ( .not. module_initialized ) call initialize_module
 
+call validate_obs_type_index(obs_type_ind, routine)
+
 use_ext_prior_this_type_of_obs = obs_type_info(obs_type_ind)%use_precomputed_FO
 
 end function use_ext_prior_this_type_of_obs
 
 !----------------------------------------------------------------------------
+!> Returns the associated generic quantity associated with the
+!> specific type of this observation
 
 function get_quantity_for_type_of_obs(obs_type_ind)
 
-! Returns the associated generic kind associated with the
-! specific type of this observation
-
 integer, intent(in) :: obs_type_ind
 integer             :: get_quantity_for_type_of_obs
 
+character(len=*), parameter :: routine = 'get_quantity_for_type_of_obs'
+
 if ( .not. module_initialized ) call initialize_module
 
-get_quantity_for_type_of_obs = obs_type_info(obs_type_ind)%generic_kind
+call validate_obs_type_index(obs_type_ind, routine)
+
+get_quantity_for_type_of_obs = obs_type_info(obs_type_ind)%quantity
 
 end function get_quantity_for_type_of_obs
 
 !----------------------------------------------------------------------------
+!> Writes out the observation type strings and corresponding integer
+!> indices as a header for an obs_sequence file.
 
 subroutine write_type_of_obs_table(ifile, fform, use_list)
 
-! Writes out the observation kind strings and corresponding integer
-! indices as a header for an obs_sequence file.
-
 integer, intent(in)                    :: ifile
 character(len=*), intent(in), optional :: fform
 integer,          intent(in), optional :: use_list(:)
 
 integer :: i, ntypes
 logical :: is_ascii, restrict
+character(len=*), parameter :: routine = 'write_type_of_obs_table'
 
 if ( .not. module_initialized ) call initialize_module
 
 is_ascii = ascii_file_format(fform)
 
-! Write the 20 character identifier to indicate the start of a kind table.
+! Write the 20 character identifier to indicate the start of a type table.
 ! To maintain complete backwards compatibility with the original read
 ! code this string must start in column 1.
 ! you cannot use: write(ifile, *) 'string'  because the output string
 ! will start in col 2.  the string either has to be part of the format
 ! argument to write(), or you have to use a separate format statement.
+! FIXME: this needs to be obs_type_definitions for write.  read has to
```

Comment:
this isn't a FIXME anymore.  this code *is* writing out 'obs_type_definitions' in newly created obs_seq files, and the read routine is accepting either the new or old string so it's fully backwards compatible.  so the review input is this comment should be changed from 'should' or 'needs to' into something that has already changed.